barretenberg/cpp/src/barretenberg/ecc/fields/field_conversion_utils.hpp

```diff
@@ -0,0 +1,387 @@
+#pragma once
+
+#include "barretenberg/ecc/curves/bn254/bn254.hpp"
+#include "barretenberg/ecc/curves/bn254/fr.hpp"
+#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
+#include "barretenberg/plonk/proof_system/constants.hpp"
+#include "barretenberg/polynomials/univariate.hpp"
+
+namespace barretenberg::field_conversion_utils {
+
+/**
+ * @brief Decomposes a barretenberg::fr into two 64-bit limbs. Helper function for
+ * convert_barretenberg_frs_to_grumpkin_fr.
+ *
+ * @param field_val
+ * @return std::array<uint64_t, 2>
+ */
+std::array<uint64_t, 2> decompose_bn254_fr_to_two_limbs(const barretenberg::fr& field_val);
+
+/**
+ * @brief Converts 2 barretenberg::fr elements to grumpkin::fr
+ * @details Checks that each barretenberg::fr must be at most 128 bits (to ensure no overflow), and decomposes each
+ * barretenberg::fr into two 64-bit limbs, and the 4 64-bit limbs form the grumpkin::fr
+ * @param low_bits_in
+ * @param high_bits_in
+ * @return grumpkin::fr
+ */
+grumpkin::fr convert_bn254_frs_to_grumpkin_fr(const barretenberg::fr& low_bits_in,
+                                              const barretenberg::fr& high_bits_in);
+
+/**
+ * @brief Converts grumpkin::fr to 2 barretenberg::fr elements
+ * @details Does the reverse of convert_bn254_frs_to_grumpkin_fr, by merging the two pairs of limbs back into the
+ * 2 barretenberg::fr elements.
+ * @param input
+ * @return std::array<barretenberg::fr, 2>
+ */
+std::array<barretenberg::fr, 2> convert_grumpkin_fr_to_bn254_frs(const grumpkin::fr& input);
+
+/**
+ * @brief Calculates the size of a types in terms of barretenberg::frs
+ * @details We want to suppor the following types: bool, size_t, uint32_t, uint64_t, barretenberg::fr, grumpkin::fr,
+ * curve::BN254::AffineElement, curve::Grumpkin::AffineElement, barretenberg::Univariate<FF, N>, std::array<FF, N>, for
+ * FF = barretenberg::fr/grumpkin::fr, and N is arbitrary
+ * @tparam T
+ * @return constexpr size_t
+ */
+template <typename T> constexpr size_t calc_num_frs();
+
+constexpr size_t calc_num_frs(barretenberg::fr* /*unused*/)
+{
+    return 1;
+}
+
+constexpr size_t calc_num_frs(grumpkin::fr* /*unused*/)
+{
+    return 2;
+}
+
+template <std::integral T>
+constexpr size_t calc_num_frs(T* /*unused*/) // TODO: check if std integral includes uint256 and uint512, those are too
```

Comment:
integral does not include uint256 though it might include a compiler-provided uint128 i think